File: golang/array/剑指Offer12. 矩阵中的路径.go
package array

var (
	dx = []int{-1, 1, 0, 0}
	dy = []int{0, 0, -1, 1}
)

func exist(board [][]byte, word string) bool {
	m := make([][]int, 8)
	for i := 0; i < 8; i++ {
		m[i] = make([]int, 8)
	}

	for i := 0; i < len(board); i++ {
		for j := 0; j < len(board[i]); j++ {
			m[i][j] = 1
			res := wordFind(m, board, i, j, 0, word)
			if res {
				return true
			}
			m[i][j] = 0
		}
	}

	return false
}

func wordFind(m [][]int, board [][]byte, i, j, now int, word string) bool {
	if board[i][j] != word[now] {
		return false
	}
	if now == len(word)-1 {
		return true
	}

	for k := 0; k < 4; k++ {
		nx, ny := i+dx[k], j+dy[k]
		if nx < 0 || nx >= len(board) || ny < 0 || ny >= len(board[0]) || m[nx][ny] == 1 {
			continue
		}
		m[nx][ny] = 1
		res := wordFind(m, board, nx, ny, now+1, word)
		if res {
			return res
		}
		m[nx][ny] = 0
	}

	return false
}

File: java-ladidol/src/com/feng/leetcode/offer/_07搜索和回溯/剑指Offer12矩阵中的路径.java
package com.feng.leetcode.offer._07搜索和回溯;

/**
 * @author: Xiaoqiang-Ladidol
 * @date: 2023/7/24 21:54
 * @description: {}
 */
public class 剑指Offer12矩阵中的路径 {




    //dfs+回溯
    class Solution {
        /**
         * 根据首字母寻找可能的情况，不断深度往下。
         * 注意回溯：使用了棋盘也要复原
         * @param board
         * @param word
         * @return
         */
        public boolean exist(char[][] board, String word) {

            int n = board.length;
            int m = board[0].length;
            char[] words = word.toCharArray();

//                visited = new int[n][m];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (flag) return true;
                    dfs(board, i, j, 0, words);
                }
            }

            return flag;
        }

        int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        boolean flag = false;
//            int[][] visited;

        void dfs(char[][] board, int i, int j, int index, char[] words) {
            if (board[i][j] != words[index]) {
                return;
            }
            if (index == words.length - 1) {
                flag = true;
                return;
            }
            char t = board[i][j];
            board[i][j] = '#';//被搜索过。

            for (int k = 0; k < 4; k++) {
                int x = i + dirs[k][0];
                int y = j + dirs[k][1];
                if (x >= 0 && x < board.length && y >= 0 && y < board[0].length)
                    dfs(board, x, y, index + 1, words);
            }
            //注意回溯
            board[i][j] = t;//搜索状态还原。

        }
    }



}
